File: Monads/Monaden Solution/Monaden/State.fs
namespace Monaden

open System

module State =

    type StateM<'s,'a> = internal { runState : 's -> ('a*'s) }

    let mkState f = { runState = f }

    let returnM (a : 'a) : StateM<'s,'a> =
        mkState <| fun s -> (a, s)

    let bind (f : StateM<'s,'a>) (g : 'a -> StateM<'s,'b>) : StateM<'s,'b> =
        fun s ->
            let (a,s') = f.runState s
            (g a).runState s'
        |> mkState

    let (>>=) m f =
        bind m f

    let (>=>) f g =
        fun a -> f a >>= g

    type StateBuilder internal () =
        member x.Bind(m, f) = m >>= f
        member x.Return(v) = returnM v
        member x.ReturnFrom(v) = v
        member x.Delay(f) = f ()

    let state = StateBuilder()    

    let eval (s : 's) (m : StateM<'s,'a>) : 'a =
        m.runState s
        |> fst

    let get() : StateM<'s,'s> =
        fun s -> (s,s)
        |> mkState

    let set (s : 's) : StateM<'s, unit> =
        fun _ -> ((), s)
        |> mkState

    module Beispiel =
        
        let nextFib : StateM<int*int, int> =
            state {
                let! l1,l2 = get()
                let akt = l1+l2
                do! set (l2,akt)
                return akt
            }

        let test () : int =
            state {
                let! _ = nextFib
                let! _ = nextFib
                return! nextFib
            } |> eval (1,1) // = 5

File: Monads/Monaden Solution/Monaden.CSharp/State/State.cs
using System;

namespace Monaden.CSharp.State
{
    public class StateM<S,T>
    {
        readonly Func<S, Tuple<T, S>> _runState;

        internal StateM(Func<S,Tuple<T,S>> runState)
        {
            _runState = runState;
        }

        public Tuple<T,S> Run(S state)
        {
            return _runState(state);
        }

        public T Eval(S state)
        {
            return Run(state).Item1;
        }

        public S Process(S state)
        {
            return Run(state).Item2;
        }

        internal StateM<S,R> BindMap<C,R>(Func<T,StateM<S,C>> f, Func<T,C,R> map)
        {
            return new StateM<S, R>(s =>
            {
                var t1 = _runState(s);
                var m2 = f(t1.Item1);
                var t2 = m2.Run(t1.Item2);
                return Tuple.Create(map(t1.Item1, t2.Item1), t2.Item2);
            });
        }

        internal StateM<S, R> Bind<R>(Func<T, StateM<S, R>> f)
        {
            return BindMap(f, (_, x) => x);
        }

        internal StateM<S,R> Map<R>(Func<T, R> map)
        {
            return new StateM<S,R>(s => {
                var t = Run(s);
                return Tuple.Create(map(t.Item1), t.Item2);
            });
        }

    }

    public static class StateM
    {
        public static StateM<S, T> Return<S, T>(this T value)
        {
            return new StateM<S, T>(s => Tuple.Create(value, s));
        }

        public static StateM<S,R> Select<S,T,R>(this StateM<S,T> source, Func<T,R> map)
        {
            return source.Map(map);
        }

        public static StateM<S,B> SelectMany<S, A, B>(this StateM<S, A> source, Func<A, StateM<S, B>> selector)
        {
            return source.Bind(selector);
        }

        public static StateM<S,C> SelectMany<S, A, B, C>(this StateM<S, A> source, Func<A, StateM<S, B>> collectionSelector, Func<A, B, C> resultSelector)
        {
            return source.BindMap(collectionSelector, resultSelector);

        }

        public static StateM<S, S> Get<S>()
        {
            return new StateM<S, S>(s => Tuple.Create(s, s));
        }

        public static StateM<S, Unit> Set<S>(this S state)
        {
            return new StateM<S, Unit>(_ => Tuple.Create(Unit.V, state));
        }

        public static Func<A, StateM<S, C>> Comp<S, A, B, C>(this Func<A, StateM<S, B>> f, Func<B, StateM<S, C>> g)
        {
            return a => f(a).Bind(g);
        }

    }
}
